pkgs/top-level/all-packages.nix

```diff
@@ -14815,22 +14815,25 @@ in
 
   boolstuff = callPackage ../development/libraries/boolstuff { };
 
-  boost155 = callPackage ../development/libraries/boost/1.55.nix { };
-  boost159 = callPackage ../development/libraries/boost/1.59.nix { };
+  inherit (callPackage ../development/libraries/boost { inherit (buildPackages) boost-build; })
```

Comment:
```suggestion
  inherit (callPackages ../development/libraries/boost { inherit (buildPackages) boost-build; })
```

Reply:
Thanks!

Reply:
Actually had to revert that, since `callPackages` messes with the `override` of the inner attributes. It's fine how it is currently since we `callPackage` the derivations.

Reply:
ah I see, that makes sense.

Reply:
It needs to mess with each package to do the splicing magic for cross compilation; the machinery that lets you use the same variable to represent multiple possible derivations, a different one of which is used in each of `buildInputs`, `nativeBuildInputs`, `deps${x}${y}`.

If you do this one level too soon, as you do here, the decision is made too soon, resulting in hard to understand cross breakage.

Reply:
So how do I best put this in another file? `import`? Not sure what you are reffering to exactly, tbh.

Reply:
I probably did a terrible job trying to explain it. Thing is, it's underdocumented. The manual has this line

> To make this work, we “splice” together the six pkgsFooBar package sets and have callPackage actually take its arguments from that.

Which is hidden away in a cross-compilation section. This should also be part of the `callPackage` doc. Oh, which is never introduced.

I don't understand why it doesn't work for you. Maybe @Ericson2314 has experienced this?

`import` and many `callPackage` calls should work like it did before. That could be a way out.

Reply:
It works for me at his, I thought you were implying using `callPackage` in the way I am would be a problem. So far I have seen no evidence of that, e. g. `pkgsLLVM.nix` builds fine which does rely on cross-compilation.

Reply:
Maybe it still works because boost is always used in `buildInputs`, which is the one attribute where a spliced package produces the same result as a non-spliced one.
It will break if it occurs as a dependency with a different [role](https://nixos.org/manual/nixpkgs/unstable/#variables-specifying-dependencies). This is extremely rare for boost, if not always wrong. I've found one weird example, `pkgs/applications/science/electronics/horizon-eda/default.nix` but I doubt its correctness. One example after looking way too hard, like 40% of usages. It might be ok to break.

Reply:
Oh yeah, indeed. It does bug me a bit though, will try to find a way to solve this.

Reply:
Nevermind, this wasn't even working before: In `nixos-unstable`:

```
nix-repl> callPackage ({ boost }: boost ? __spliced && boost ? nativeDrv && boost ? crossDrv) { }
false
```

Maybe something for a future refactor.

Reply:
These are only present when `pkgs` is cross.

```
nix-repl> pkgs = import ./. { crossSystem = "aarch64-linux"; localSystem = "x86_64-linux"; }          

nix-repl> pkgs.callPackage ({ boost }: boost?__spliced && boost?crossDrv && boost?nativeDrv) {}
true

```

So it does work.

Reply:
Of course, yeah, can' think anymore today. On the plus side, it also seems to work with this PR applied, so that's good news!

Reply:
Isn't `inherit (buildPackages) boost-build;` effectively implied when you're only using it in `nativeBuildInputs`?

Reply:
Yes, but since we `override` the `boost-build` derivation later, the splicing magic won't be able to process it and pass it through as is.

I think it is much nicer to use `buildPackages` explicitly here to work around it than to do `boost-build.nativeDrv.override`…

Reply:
I'm not convinced that splicing can't work with `override` yet, so I've split this off into #132651, which is in no way a blocker of course.

Reply:
I also noticed that the splicing logic traverses non-drv attrsets, so that explains why it still works.

Reply:
I think it works because every derivation we inherit is a result of three nested `callPackage` calls (lol).